packages/lab/src/carousel/CarouselContext.tsx

```diff
@@ -0,0 +1,154 @@
+import { createContext, useResizeObserver } from "@salt-ds/core";
+import {
+  type KeyboardEvent,
+  type MouseEvent,
+  type ReactNode,
+  type RefObject,
+  useCallback,
+  useContext,
+  useEffect,
+  useRef,
+  useState,
+} from "react";
+
+export interface CarouselContextValue {
+  firstVisibleSlide: number;
+  visibleSlides: number;
+  slidesCount: number;
+  nextSlide: (event: MouseEvent | KeyboardEvent) => void;
+  prevSlide: (event: MouseEvent | KeyboardEvent) => void;
+  updateActiveFromScroll: (scrollLeft: number) => void;
+  focusSlide: (index: number) => void;
+  registerSlide: (element: HTMLDivElement) => number;
+  unregisterSlide: (index: number) => void;
+  containerRef: RefObject<HTMLDivElement>;
+  carouselId?: string;
+}
+
+export const CarouselContext = createContext<CarouselContextValue | null>(
+  "CarouselContext",
+  null,
+);
+
+export function useCarousel() {
+  const context = useContext(CarouselContext);
+  if (!context) {
+    throw new Error("useCarousel must be used within carousel provider");
+  }
+  return context;
+}
+
+export function CarouselProvider({
+  children,
+  activeSlideIndex = 0,
+  visibleSlides = 1,
+  id,
+}: {
+  children: ReactNode;
+  activeSlideIndex?: number;
+  visibleSlides?: number;
+  id?: string;
+}) {
+  const [firstVisibleSlide, setFirstVisibleSlide] = useState(activeSlideIndex);
+  const [visibleFocus, setVisibleFocus] = useState(0);
+  const [sliderW, setSliderW] = useState(0);
+  const slides = useRef<Map<number, HTMLDivElement>>(new Map());
+  const [slidesCount, setSlidesCount] = useState(slides.current.size);
```

Comment:
Do we need to maintain this state or can we derive it from `slides.current.size` ?

Reply:
slides.current.size would not update immediately, so the updates get to the consumer components one render late if i depend on it.